Fix Perl5 GH #17318

khwilliamson opened the pull request:
See https://github.com/Perl/perl5/issues/17318

khwilliamson committed:
Fix Perl5 GH #17318

See https://github.com/Perl/perl5/issues/17318

theory commented:
Thankyou, @khwilliamson. How would you describe the change in for the `Changes` file?

theory commented:
Also, should we add some tests with valid UTF-8 uses of the CI bytes, such as π? IOW, to ensure it doesn't do this:

```
 > perl -MData::Dump -E 'dd "π"'
"\xCF\x80"
 > perl -MData::Dump -MEncode::ZapCP1252=fix_cp1252 -E 'dd fix_cp1252 "π"'
"\xCF\xE2\x82\xAC"
```

khwilliamson commented:
On 12/11/19 7:06 PM, David E. Wheeler wrote:
> Thankyou, @khwilliamson <https://github.com/khwilliamson>. How would you 
> describe the change in for the |Changes| file?
> 

How about

This module no longer creates malformed UTF-8 under conditions where it 
previously did

khwilliamson commented:
On 12/11/19 7:08 PM, David E. Wheeler wrote:
> Also, should we add some tests with valid UTF-8 uses of the CI bytes, 
> such as π? IOW, to ensure it doesn't do this:
> 
> |> perl -MData::Dump -E 'dd "π"' "\xCF\x80" > perl -MData::Dump 
> -MEncode::ZapCP1252=fix_cp1252 -E 'dd fix_cp1252 "π"' "\xCF\xE2\x82\xAC" |
> 

Sure

theory committed:
Test valid 0x80 in UTG-8.

It should not be replace, but should be replaced when the string is not Unicode.
Not sure why test 9 fails; it should be adding UTF-8 bytes to the byte
string, I think but instead adds the decoded euro symbol.

theory commented:
Okay, I modified Changes and added some tests, but I expected it to replace `"\x80"` with `"\xE2\x82\xAC"`, but it instead replaced it with `"\x{20AC}"`, I think. I'm thinking it should use encoded bytes when dealing with a non-Unicode string, no?

khwilliamson commented:
I'll try to look tomorrow

khwilliamson commented:
Could you send me the exact test case you think is wrong?

theory commented:
It's [this one](/theory/encode-zapcp1252/blob/7e42fb97369741c916b82849ad3e34954d335276/t/decoded.t#L66).

khwilliamson commented:
I was able to change things to get that to work, but then other things break.  We can't have it both ways.

Look at this test in base.t

`fix_cp1252 $fix_me;
is $fix_me, $utf8, 'Convert to utf-8';`

where the scalar is a byte string.  That is supposed to convert the result to UTF-8.
The test you added is 
`is fix_cp1252 "\xCF\x80", "\xCF" . Encode::encode_utf8($euro),`

and you don't want it to be UTF-8.  There is no fundamental difference in the inputs, hence the expectations of the result should be the same as well.

theory commented:
It has been long enough now that I've lost the thread of it. Did I really make it so that `fix_cp1252`, when it encountered a string with the `utf8` flag turned off, it would replace the CP1252 bytes with Unicode code points (`"\x{20AC}"`) instead of UTF-8 bytes (`"\xE2\x82\xAC"`)? That doesn't seem like a good idea, does it?

theory commented:
IOW, would this not be more sensible?

``` patch
diff --git a/lib/Encode/ZapCP1252.pm b/lib/Encode/ZapCP1252.pm
index a3f24c4..cd067a5 100644
--- a/lib/Encode/ZapCP1252.pm
+++ b/lib/Encode/ZapCP1252.pm
@@ -137,12 +137,14 @@ sub _tweakit {
     local $_[0] = $_[0] if defined wantarray;
     my $is_utf8 = PERL588 && Encode::is_utf8($_[0]);
     my $valid_utf8 = $is_utf8 && utf8::valid($_[0]);
-    if ($valid_utf8 || ! $is_utf8) {
+    if (!$is_utf8) {
 
         # Here is either not UTF-8, or the UTF-8 is valid.  Change the 1252
         # characters to their counterparts.  (These bytes are very rarely used
         # in real world applications, so their presence likely indicates that
         # CP1252 was meant, regardless of whether the string is UTF-8 or not.)
+        $_[0] =~ s/($cp1252_re)/Encode::encode_utf8($table->{$1})/gems;
+    } elsif ($valid_utf8) {
         $_[0] =~ s/($cp1252_re)/$table->{$1}/gems;
     } else {    # Invalid UTF-8.  Look for single-byte CP1252 gremlins
 
```

theory commented:
Originally, the tests didn't have `use utf8`. This works:

``` patch
diff --git a/lib/Encode/ZapCP1252.pm b/lib/Encode/ZapCP1252.pm
index a3f24c4..cd067a5 100644
--- a/lib/Encode/ZapCP1252.pm
+++ b/lib/Encode/ZapCP1252.pm
@@ -137,12 +137,14 @@ sub _tweakit {
     local $_[0] = $_[0] if defined wantarray;
     my $is_utf8 = PERL588 && Encode::is_utf8($_[0]);
     my $valid_utf8 = $is_utf8 && utf8::valid($_[0]);
-    if ($valid_utf8 || ! $is_utf8) {
+    if (!$is_utf8) {
 
         # Here is either not UTF-8, or the UTF-8 is valid.  Change the 1252
         # characters to their counterparts.  (These bytes are very rarely used
         # in real world applications, so their presence likely indicates that
         # CP1252 was meant, regardless of whether the string is UTF-8 or not.)
+        $_[0] =~ s/($cp1252_re)/Encode::encode_utf8($table->{$1})/gems;
+    } elsif ($valid_utf8) {
         $_[0] =~ s/($cp1252_re)/$table->{$1}/gems;
     } else {    # Invalid UTF-8.  Look for single-byte CP1252 gremlins
 
diff --git a/t/base.t b/t/base.t
index 27bac47..b204e8b 100755
--- a/t/base.t
+++ b/t/base.t
@@ -1,7 +1,7 @@
 #!/usr/bin/perl -w
 
 use strict;
-use utf8;
+#use utf8;
 use Test::More tests => 13;
 
 BEGIN { use_ok 'Encode::ZapCP1252' or die; }
diff --git a/t/perl-510.t b/t/perl-510.t
index 00d5f10..a009a3c 100644
--- a/t/perl-510.t
+++ b/t/perl-510.t
@@ -1,7 +1,7 @@
 #!/usr/bin/perl -w
 
 use strict;
-use utf8;
+#use utf8;
 use Test::More;
 
 BEGIN {
```

theory committed:
Don't insert Unicode code points when utf8 is off.

Also fix the link to Encoding::FixLatin.

theory commented:
Pushed the fix in 411b0a6. I think it's correct to insert UTF-8 bytes when the utf8 flag isn't set, and unicode code points when it is. Does that seem right to you, @khwilliamson?

theory committed:
Extend comments.

theory committed:
Juse use Encode.

theory committed:
Remove `use utf8`.

This module is optimized for handling bytes; processing encoded strings should
be the exception, so make it the secondary use case.

theory committed:
Discourage use of Unicode strings

Use UTF-8 bytes, and fall back on handling encoded strings only as
a last resort.

theory commented:
Hey @khwilliamson, I'm inclined to cut a release of 411cb40. Any objections to or errors in the changes I've made since you pushed a6f3654?

khwilliamson commented on lib/Encode/ZapCP1252.pm:
```diff
@@ -73,40 +73,135 @@ our %utf8_for = (
+        # counterparts. (These bytes are very rarely used in real world
+        # applications, so their presence likely indicates that CP1252 was
+        # meant, regardless of whether the string is UTF-8 or not.)
+        $_[0] =~ s/($cp1252_re)/$table->{$1}/gems;
```
I don't understand.  One of your commits wrapped $table with encode_utf8(), and it created a compilation error.  But I think that was the right thing to do, and I don't see how it got removed, unless github did it manually

theory replied:
I removed `utf8` and switched things to call `decode_utf8()` where appropriate, instead, because the module is mainly intended for dealing with bytes, with Unicode string handling as a secondary, less recommended use case.

khwilliamson commented on lib/Encode/ZapCP1252.pm:
```diff
@@ -73,40 +73,135 @@ our %utf8_for = (
+        # Here is non-UTF-8. Change the 1252 characters to their UTF-8
+        # counterparts. (These bytes are very rarely used in real world
+        # applications, so their presence likely indicates that CP1252 was
+        # meant, regardless of whether the string is UTF-8 or not.)
```
That 'regardless' is no longer applicable because you've split the actions.

theory replied:
Thanks, fixed in 608dbb7.

khwilliamson commented on lib/Encode/ZapCP1252.pm:
```diff
@@ -73,40 +73,135 @@ our %utf8_for = (
+        $_[0] =~ s/($cp1252_re)/$table->{$1}/gems;
+    } elsif ($valid_utf8) {
+
+        # Here is well-formed Perl extended UTF-8 and has the UTF-8 flag on
```
I saw a diff where this was spelled 'Heere', and it seems to have gotten corrected, but I know not how

theory replied:
Yeah, I fixed it in 06e89e9.

khwilliamson commented on t/base.t:
```diff
@@ -1,6 +1,7 @@
 #!/usr/bin/perl -w
 
 use strict;
+#use utf8;
```
I presume you would just delete this line instead of commenting out; likewise in another spot

theory replied:
Yes, mainly left it until you told me it wasn't dumb. :-)

theory replied:
Removed in 608dbb7.

khwilliamson reviewed (commented):
If this is passing tests, I have no problem with it.

theory committed:
Fix comments.
